src/bd/primitives/Specializer.java

```diff
@@ -0,0 +1,118 @@
+package bd.primitives;
+
+import java.lang.reflect.InvocationTargetException;
+
+import com.oracle.truffle.api.dsl.NodeFactory;
+import com.oracle.truffle.api.source.SourceSection;
+
+import bd.primitives.Primitive.NoChild;
+import bd.settings.VmSettings;
+
+
+/**
+ * A Specializer defines when a node can be used as a eager primitive, how
+ * it is to be instantiated, and acts as factory for them.
+ *
+ * @param <T> the type of the nodes created by this specializer
+ * @param <Context> the type of the context object
+ * @param <ExprT> the root type of expressions used by the language
+ */
+public class Specializer<T, Context, ExprT> {
+  protected final Context                    context;
```

Comment:
The Smalltalk term `context` is non-standard compared to the rest of computer science. I'd rather go with `frame` instead of Smalltalk's `context` for Smalltalk, too.
And for BD, I'd rather stick with Truffle terminology. I'd think it makes it easier to match up concepts, since Truffle is a definite requirement anyway.